### PR TITLE
plan: fix panic caused by empty schema of LogicalTableDual

### LINKED ISSUE #7833: DROP VIEW IF EXISTS causes server crash (breaking mydumper)
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?

myloader tries to drop a table as both a view and a table when restoring with the `-o` option (overwrite).  It crashes tidb, which I managed to reduce to this test case:

```
MySQL [test]> CREATE TABLE t1 (a int);
DRQuery OK, 0 rows affected (0.53 sec)

MySQL [test]> DROP VIEW IF EXISTS t1;
ERROR 2013 (HY000): Lost connection to MySQL server during query
```

2. What did you expect to see?

Not crash.

3. What did you see instead?

See above.

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
MySQL [test]> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v2.1.0-rc.3-5-g99c4a15ce
Git Commit Hash: 99c4a15ce422603a6ad02be8ace19a43f8a26162
Git Branch: newmaster
UTC Build Time: 2018-10-07 04:24:32
GoVersion: go version go1.11 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by tianjiqx**
### What problem does this PR solve? 

fix #7833 

### What is changed and how it works?

Initialize null pointer
### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by eurekaka** [comment]
I thinks we should not enforce `LogicalTableDual` to have a valid schema, because semantically speaking, it has no such constraint, instead I prefer to add a null check in `LogicalTableDual::PruneColumns` for `p.schema`, like say:
```go
func (p *LogicalTableDual) PruneColumns(parentUsedCols []*expression.Column) {
    if p.schema == nil {
        return
    }
    xxx
}
```
If we choose to enforce valid schema of `LogicalTableDual`, we have to double check all existing `LogicalTableDual` in current code base then.

File: planner/core/preprocess_test.go [hunk 1]
```diff
@@ -194,6 +194,9 @@ func (s *testValidatorSuite) TestValidator(c *C) {
 		{"select * from ( select 1 ) a, (select 2) b;", true, nil},
 		{"select * from (select * from ( select 1 ) a join (select 2) b) b join (select 3) a;", false, nil},
 		{"select * from (select 1 ) a , (select 2) b, (select * from (select 3) a join (select 4) b) c;", false, nil},
+
+		//issue 7833
+		{"drop view if exists v;", true, nil},
```
Comment (eurekaka) [inline planner/core/preprocess_test.go]: Better to add this test in `TestColumnPruning`.

**Comment by eurekaka** [comment]
Kind reminder, this PR need to be cherry-picked into release-2.1 branch after merged into master. release-2.0 branch does not have this problem.

File: planner/core/planbuilder.go [hunk 2]
```diff
@@ -227,6 +227,8 @@ func (b *PlanBuilder) buildDo(v *ast.DoStmt) (Plan, error) {
 			RetType:  expr.GetType(),
 		})
 	}
+	//make dual schema if not
+	var _ = dual.Schema()
```
Comment (zz-jason) [inline planner/core/planbuilder.go]: how about directly set the schema of the `dual` operator, which eliminates the confusion:
```go
dual.schema = expression.NewSchema()
```

Comment (tianjiqx) [reply planner/core/planbuilder.go]: Yes, it looks better

File: planner/core/preprocess_test.go (hunk 1 above)
Comment (tianjiqx) [reply planner/core/preprocess_test.go]: Well, I've found that this test case can be added here（planner/core/preprocess_test.go）, and even without fixing the bug, I'm confused, maybe I don't fully understand the tidb test.

**Comment by zz-jason** [comment]
@eurekaka @lamxTyler @lzmhhh123 PTAL

**Pull request merged**